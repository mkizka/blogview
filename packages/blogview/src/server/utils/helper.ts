import fs from "fs";
import path from "path";
import { fileURLToPath } from "node:url";

export function loadJson(filepath: string) {
  try {
    const text = fs.readFileSync(filepath, "utf-8");
    return JSON.parse(text) as Record<string, any>;
  } catch (e) {
    return null;
  }
}

const dirname = path.dirname(fileURLToPath(import.meta.url));

const pkgRootDir = path.join(dirname, "..", "..", "..");

export function pkgPath(filepath: string) {
  return path.join(pkgRootDir, filepath);
}
