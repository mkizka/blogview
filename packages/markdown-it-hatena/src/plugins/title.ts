import { PluginWithOptions } from "markdown-it";
import markdownItFrontMatter from "markdown-it-front-matter";
import yaml from "js-yaml";

export type BlogMeta = {
  categories?: string[];
  date?: Date;
  draft?: boolean;
  id?: string;
  title?: string;
};

export type TitlePluginOptions = {
  hideTitle?: boolean;
  cb?: (meta: unknown) => void;
};

function noop() {}

function date(date: Date) {
  return `
  <div class="date entry-date first">
    <a href="" rel="nofollow">
      <time datetime="${date.toJSON()}" title="${date.toJSON()}">
        <span class="date-year">${date.getFullYear()}</span>
        <span class="hyphen">-</span>
        <span class="date-month">${date.getMonth() + 1}</span>
        <span class="hyphen">-</span>
        <span class="date-day">${date.getDate()}</span>
      </time>
    </a>
  </div>
`.trim();
}

function title(title: string) {
  return `
  <h1 class="entry-title">
    <a href="" class="entry-title-link bookmark">${title}</a>
  </h1>
`.trim();
}

function categories(categories: string[]) {
  return `
  <div class="entry-categories categories">
    ${categories
      .map(
        (category: string) =>
          `<a href="" class="entry-category-link category-${category}">${category}</a>`
      )
      .join("\n    ")}
  </div>
`.trim();
}

export const titlePlugin: PluginWithOptions<TitlePluginOptions> = (
  md,
  options
) => {
  const cb = options?.cb || noop;
  md.use(markdownItFrontMatter, (metaRaw) => cb(yaml.load(metaRaw)));
  if (options?.hideTitle) return;
  md.renderer.rules.front_matter = (tokens, idx) => {
    const meta = yaml.load(tokens[idx].meta) as BlogMeta;
    return `
<header class="entry-header">
  ${[
    meta.date && date(meta.date),
    meta.title && title(meta.title),
    meta.categories && categories(meta.categories),
  ]
    .filter(Boolean)
    .join("\n  ")}
</header>
`.trim();
  };
};
