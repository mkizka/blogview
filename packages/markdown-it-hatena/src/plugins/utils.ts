import UrlRegex from "url-regex";

const urlRegex = new RegExp(UrlRegex().source, "i");

type HatenaNotationBase = {
  type: string;
  src: string;
};

export type HatenaNotationLink = HatenaNotationBase & {
  type: "link";
  url: string;
  options: string[];
};

export type PlainText = HatenaNotationBase & {
  type: "text";
};

export type HatenaNotation = HatenaNotationLink | PlainText;

export function parseHatenaNotation(text: string): HatenaNotation[] {
  const result = [""];
  for (const char of text) {
    switch (char) {
      case "[":
        result.push(char);
        break;
      case "]":
        result[result.length - 1] += char;
        result.push("");
        break;
      default:
        result[result.length - 1] += char;
        break;
    }
  }
  return result.filter(Boolean).map(parse);
}

function parse(src: string): HatenaNotation {
  // [ ]で囲まれているもののみはてな記法として扱う
  if (src.startsWith("[") && src.endsWith("]")) {
    const innerText = src.slice(1, src.length - 1);
    const splitted = innerText.split(":");
    // https://help.hatenablog.com/entry/editor/advlink
    const maybeUrl = splitted.slice(0, 2).join(":"); // "https://"が区切られてしまっているため
    if (urlRegex.test(maybeUrl)) {
      return {
        type: "link",
        url: maybeUrl,
        options: splitted.slice(2),
        src,
      };
    }
  }
  return {
    type: "text",
    src,
  };
}
