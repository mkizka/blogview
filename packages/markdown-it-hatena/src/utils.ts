export function parseHatenaNotation(text: string) {
  const matched = /^\[(?<url>.*?)(?<labels>(:\w+)*)\]$/g.exec(text);
  if (matched) {
    return {
      url: matched.groups!.url,
      labels: matched.groups!.labels.split(":").filter(Boolean),
    };
  }
  return null;
}
