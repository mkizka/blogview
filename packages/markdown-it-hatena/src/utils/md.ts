import MarkdownIt from "markdown-it";

export function createMarkdownIt() {
  return new MarkdownIt({ linkify: true, html: true });
}
