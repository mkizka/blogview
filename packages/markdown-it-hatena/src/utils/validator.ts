export function isTwitter(url: string) {
  return /^https:\/\/twitter\.com\/\S+?\/status/.test(url);
}

export function isYouTube(url: string) {
  return /^https:\/\/www\.youtube\.com\/watch/.test(url);
}
