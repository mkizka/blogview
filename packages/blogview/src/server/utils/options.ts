import path from "path";
import arg from "arg";

import { loadJson } from "./helper.js";

export const optionsDefault = {
  entry: "entry",
  styles: [] as string[],
  twitter: true,
  youtube: true,
};

function removeUndefinedField(obj: unknown) {
  return JSON.parse(JSON.stringify(obj)) as Record<string, any>;
}

export type Options = typeof optionsDefault;

export function getOptions(argv?: string[]) {
  const args = arg(
    {
      "--config": String,
      "-c": "--config",
      "--entry": String,
      "-e": "--entry",
      "--style": [String],
      "--no-twitter": Boolean,
      "--no-youtube": Boolean,
    },
    { permissive: true, argv }
  );

  const optionsFromArgs = removeUndefinedField({
    entry: args["--entry"],
    styles: args["--style"],
    twitter: args["--no-twitter"] && false,
    youtube: args["--no-youtube"] && false,
  });

  const optionsFromConfig = loadJson(
    path.join(process.cwd(), args["--config"] || "blogview.json")
  );

  return {
    ...optionsDefault,
    ...optionsFromConfig,
    ...optionsFromArgs,
  };
}
