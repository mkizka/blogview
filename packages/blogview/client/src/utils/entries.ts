function getEntryId(path: string) {
  return path.replace("../../entry/", "").replace(".md", "");
}

export const entries = Object.entries(
  import.meta.globEager("../../entry/*.md")
).reduce((result, [path, { default: entry }]) => {
  const entryId = getEntryId(path);
  result[entryId] = entry;
  return result;
}, {} as { [entryId: string]: string });
