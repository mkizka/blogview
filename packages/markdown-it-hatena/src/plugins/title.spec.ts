import { test, expect } from "vitest";

import { titlePlugin } from "./title";
import { createMarkdownIt } from "../utils/md";

const md = createMarkdownIt().use(titlePlugin);

const title = "title: タイトル";
const date = "date: 2021-12-30T15:32:03.000Z";
const categories = "categories: [カテゴリ1, カテゴリ2]";

test("タイトルを描画", () => {
  const rendered = md.render(`---\n${title}\n---`);
  expect(rendered).toMatchSnapshot();
});

test("日付を描画", () => {
  const rendered = md.render(`---\n${date}\n---`);
  expect(rendered).toMatchSnapshot();
});

test("カテゴリを描画", () => {
  const rendered = md.render(`---\n${categories}\n---`);
  expect(rendered).toMatchSnapshot();
});

test("ヘッダ全て描画", () => {
  const rendered = md.render(`---\n${title}\n${date}\n${categories}\n---`);
  expect(rendered).toMatchSnapshot();
});
