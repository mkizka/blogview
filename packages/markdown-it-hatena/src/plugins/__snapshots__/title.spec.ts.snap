// Vitest Snapshot v1

exports[`カテゴリを描画 1`] = `
"<header class=\\"entry-header\\">
  <div class=\\"entry-categories categories\\">
    <a href=\\"\\" class=\\"entry-category-link category-カテゴリ1\\">カテゴリ1</a>
    <a href=\\"\\" class=\\"entry-category-link category-カテゴリ2\\">カテゴリ2</a>
  </div>
</header>"
`;

exports[`タイトルを描画 1`] = `
"<header class=\\"entry-header\\">
  <h1 class=\\"entry-title\\">
    <a href=\\"\\" class=\\"entry-title-link bookmark\\">タイトル</a>
  </h1>
</header>"
`;

exports[`ヘッダ全て描画 1`] = `
"<header class=\\"entry-header\\">
  <div class=\\"date entry-date first\\">
    <a href=\\"\\" rel=\\"nofollow\\">
      <time datetime=\\"2021-12-30T15:32:03.000Z\\" title=\\"2021-12-30T15:32:03.000Z\\">
        <span class=\\"date-year\\">2021</span>
        <span class=\\"hyphen\\">-</span>
        <span class=\\"date-month\\">12</span>
        <span class=\\"hyphen\\">-</span>
        <span class=\\"date-day\\">31</span>
      </time>
    </a>
  </div>
  <h1 class=\\"entry-title\\">
    <a href=\\"\\" class=\\"entry-title-link bookmark\\">タイトル</a>
  </h1>
  <div class=\\"entry-categories categories\\">
    <a href=\\"\\" class=\\"entry-category-link category-カテゴリ1\\">カテゴリ1</a>
    <a href=\\"\\" class=\\"entry-category-link category-カテゴリ2\\">カテゴリ2</a>
  </div>
</header>"
`;

exports[`日付を描画 1`] = `
"<header class=\\"entry-header\\">
  <div class=\\"date entry-date first\\">
    <a href=\\"\\" rel=\\"nofollow\\">
      <time datetime=\\"2021-12-30T15:32:03.000Z\\" title=\\"2021-12-30T15:32:03.000Z\\">
        <span class=\\"date-year\\">2021</span>
        <span class=\\"hyphen\\">-</span>
        <span class=\\"date-month\\">12</span>
        <span class=\\"hyphen\\">-</span>
        <span class=\\"date-day\\">31</span>
      </time>
    </a>
  </div>
</header>"
`;
